src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs

```diff
@@ -419,64 +419,228 @@ private void LookupMetadataDefinitions(
+                //              public static <Extension>$(parameter) { }
+                //          }
+                //
+                //          public void ExtensionMethod() { }
+                //          public int ExtensionProp => ...
+                //      }
+                //  }
+
+                if ((typeDefinition.Attributes & TypeAttributes.Sealed) == 0 ||
+                    !IsPublicType(typeDefinition) ||
+                    !HasSpecialName(typeDefinition))
+                {
+                    return null;
+                }
 
-                // We don't include internals from metadata assemblies.  It's less likely that
-                // a project would have IVT to it and so it helps us save on memory.  It also
-                // means we can avoid loading lots and lots of obfuscated code in the case the
-                // dll was obfuscated.
-                if (IsPublic(type.Attributes))
+                // Ok, we've seen the first interesting inner type.  Now look for the one inside of that
+                // that contains the maker method.
+                foreach (var nestedTypeHandle in typeDefinition.GetNestedTypes())
                 {
-                    var definition = MetadataDefinition.Create(metadataReader, type);
-                    definitionMap.Add(definition.Name, definition);
-                    _allTypeDefinitions.Add(definition);
+                    var nestedType = metadataReader.GetTypeDefinition(nestedTypeHandle);
+                    if (!IsStaticClass(nestedType) ||
+                        !IsPublicType(nestedType) ||
+                        !HasSpecialName(nestedType))
+                    {
+                        continue;
+                    }
+
+                    // Look for the special '<Extension>$' marker method which is how we can determine
+                    // which type is being extended by the extension block.
+                    foreach (var child in nestedType.GetMethods())
+                    {
+                        var method = metadataReader.GetMethodDefinition(child);
+                        if ((method.Attributes & MethodAttributes.SpecialName) == 0 &&
+                            (method.Attributes & MethodAttributes.RTSpecialName) == 0)
+                        {
+                            continue;
+                        }
+
+                        // has to be `public static "<Extension>$"(parameter)` (with exactly one parameter). 
+                        if ((method.Attributes & MethodAttributes.MemberAccessMask) != MethodAttributes.Public ||
+                            (method.Attributes & MethodAttributes.Static) == 0 ||
+                            method.GetParameters().Count != 1)
+                        {
+                            continue;
+                        }
+
+                        var methodName = metadataReader.GetString(method.Name);
+                        if (methodName != "<Extension>$")
```

Comment:
> "<Extension>$" [](http://example.com/codeflow?start=41&length=15)

Feels like this should come from `Microsoft.CodeAnalysis.WellKnownMemberNames.ExtensionMarkerMethodName` (let's make it public). Could be done in a separate PR since needs API approval

Reply:
sure.